File: library/analytics/src/debug/java/org/cru/godtools/analytics/TimberAnalyticsService.java
package org.cru.godtools.analytics;

import org.cru.godtools.analytics.model.AnalyticsActionEvent;
import org.cru.godtools.analytics.model.AnalyticsScreenEvent;
import org.cru.godtools.base.model.Event;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import timber.log.Timber;

public class TimberAnalyticsService {
    private TimberAnalyticsService() {
        EventBus.getDefault().register(this);
    }

    @Nullable
    private static TimberAnalyticsService sInstance;

    @NonNull
    public static synchronized TimberAnalyticsService start() {
        if (sInstance == null) {
            sInstance = new TimberAnalyticsService();
        }

        return sInstance;
    }

    // region Events
    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onTrackContentEvent(@NonNull final Event event) {
        Timber.tag("AnalyticsService")
                .d("onTrackContentEvent(%s:%s)", event.id.namespace, event.id.name);
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onAnalyticsScreenEvent(@NonNull final AnalyticsScreenEvent event) {
        Timber.tag("AnalyticsService")
                .d("onAnalyticsScreenEvent('%s', '%s')", event.getScreen(), event.getLocale());
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onAnalyticsActionEvent(@NonNull final AnalyticsActionEvent event) {
        Timber.tag("AnalyticsService")
                .d("onAnalyticsActionEvent('%s')", event.getAction());
    }
    // endregion Events
}

File: library/analytics/src/debug/java/org/cru/godtools/analytics/TimberAnalyticsService.kt
package org.cru.godtools.analytics

import androidx.annotation.MainThread
import javax.inject.Inject
import javax.inject.Singleton
import org.cru.godtools.analytics.model.AnalyticsActionEvent
import org.cru.godtools.analytics.model.AnalyticsScreenEvent
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import timber.log.Timber

private const val TAG = "AnalyticsService"

@Singleton
class TimberAnalyticsService @Inject internal constructor(eventBus: EventBus) {
    init {
        eventBus.register(this)
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onAnalyticsScreenEvent(event: AnalyticsScreenEvent) {
        Timber.tag(TAG)
            .d("onAnalyticsScreenEvent('%s', '%s')", event.screen, event.locale)
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onAnalyticsActionEvent(event: AnalyticsActionEvent) {
        Timber.tag(TAG)
            .d("onAnalyticsActionEvent('%s')", event.action)
    }
}
